### PR TITLE
FIPS DRBG conditional testing

### PR DESCRIPTION
**Pull request by paulidale**
Add support for the FIPS 140-2 section 4.9.2 Conditional Tests.

The code for the tests is included in all builds, but only enabled by default in FIPS builds (which currently aren't possible).  This permits any application to enable the conditional test if desired.  The new unit tests do this.

The downside of this conditional testing is that the first block of randomness read is effectively discarded.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/rand/drbg_lib.c [hunk 1]
```diff
@@ -324,10 +329,46 @@ void RAND_DRBG_free(RAND_DRBG *drbg)
...
+    }
+}
+
+/*
+ * Perform the continuous random number generator (CRNG) test.
+ * Each block of entropy is compared against the previous and the test
+ * fails if any pair is identical.  Any excess bytes are zeroed and not used.
+ * The DRBG's old entropy block is updated so the test can continue on the
+ * next call.  The block size is assumed to be non-zero, i.e. the test is
+ * executing.
```
Comment (richsalz) [inline crypto/rand/drbg_lib.c]: Consider adding "Returns 0 on failure, 1 on success" or similar.  Also, just to put it into a single comment, the function open curly needs to be on a line by itself. :)

Comment (paulidale) [reply crypto/rand/drbg_lib.c]: I only found one case of a misplaced curly (and one where I'd forgotten them).  Is there a second you noticed?

**Comment by paulidale** [comment]
Travis failure is relevant.

File: crypto/rand/drbg_lib.c (hunk 1 above)
Comment (richsalz) [reply crypto/rand/drbg_lib.c]: Just the one for the function start.

Comment (paulidale) [reply crypto/rand/drbg_lib.c]: Thanks for the confirmation.

**Comment by paulidale** [comment]
anyone?

File: crypto/rand/drbg_lib.c [hunk 2]
```diff
@@ -380,16 +426,51 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
         max_entropylen += drbg->max_noncelen;
     }
 
+    /*
+     * If we're running with continuious RNG testing and haven't initialised
```
Comment (t-j-h) [inline crypto/rand/drbg_lib.c]: "continuious" needs fixing

Comment (paulidale) [reply crypto/rand/drbg_lib.c]: Fixed.

**Comment by mspncp** [comment]
Pauli, in the commit message of 2f0067a821851eab14ef3b9461bd674298c29739 you are referring to section 4.9.2 of FIPS 140-2.

>   Refer to FIPS 140-2 section 4.9.2 Conditional Tests.
> 
>   The check is for the entropy sources to not feed the DRBG the same block of
>   seed material twice in a row.  Only the first DRBG in a chain is subject to
>   this check, latter DRBGs are assumed to be safely seeded via the earlier DRBGs.
> 
>   Outside of FIPS compiles, the test will be disabled by default but it can
>   be enabled by setting the block size.
> 

For simplicity, I am quoting the section here:

>   Continuous random number generator test. If a cryptographic module employs Approved or nonApproved RNGs in an Approved mode of operation, the module shall perform the following continuous random number generator test on each RNG that tests for failure to a constant value.
>   1. If each call to a RNG produces blocks of n bits (where n > 15), the first n-bit block generated after power-up, initialization, or reset shall not be used, but shall be saved for comparison with the next n-bit block to be generated. Each subsequent generation of an n-bit block shall be compared with the previously generated block. The test shall fail if any two compared n-bit blocks are equal.
>   2. If each call to a RNG produces fewer than 16 bits, the first n bits generated after power-up, initialization, or reset (for some n > 15) shall not be used, but shall be saved for comparison with the next n generated bits. Each subsequent generation of n bits shall be compared with the previously generated n bits. The test fails if any two compared n-bit sequences are equal. 
> 

Could you please explain why in your opinion the CRNG test needs to be applied only to the entropy _input_ of the 'master' DRBG and not to the _output_ generated by any of the DRBGs? The reason why I ask is the fact, that you are deviating from the implementation in the OpenSSL FIPS 2.0 Object Module, where the continuous RNG test was applied in the generate method (i.e., in `drbg_ctr_generate()`):

https://github.com/openssl/openssl/blob/d11e6a4410d5aae70ad545d085d344644c33d9cc/fips/rand/fips_drbg_ctr.c#L341

https://github.com/openssl/openssl/blob/d11e6a4410d5aae70ad545d085d344644c33d9cc/fips/rand/fips_drbg_lib.c#L553-L577

File: crypto/rand/drbg_lib.c [hunk 3]
```diff
@@ -380,16 +426,51 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
         max_entropylen += drbg->max_noncelen;
     }
 
+    /*
+     * If we're running with continuous RNG testing and haven't initialised
+     * yet, we need more data.  We add one block and round up to the nearest
+     * block boundary.  The entropy requirements are not altered.
```
Comment (mspncp) [inline crypto/rand/drbg_lib.c]: This comment makes me wonder whether it really makes sense to increase only the buffer size without increasing the entropy requirements. And thinking even further: does it make sense to use the "uncompressed" entropy input for the CRNG test? Shouldn't it rather be the "compressed" entropy input after applying the derivation function? (In the case where a df is used, which is our default.)

File: crypto/rand/rand_lcl.h [hunk 4]
```diff
@@ -236,6 +237,16 @@ struct rand_drbg_st {
     size_t min_noncelen, max_noncelen;
     size_t max_perslen, max_adinlen;
 
+    /*
+     * If the continuous random number testing is benig used, these hold the
```
Comment (mspncp) [inline crypto/rand/rand_lcl.h]: typo: benig -> being

**Comment by paulidale** [comment]
> Could you please explain why in your opinion the CRNG test needs to be applied only to the entropy _input_ of the 'master' DRBG and not to the _output_ generated by any of the DRBGs?

A good question.  The short answer is that I've had confirmation from the FIPS lab and our internal FIPS guy that it can be done this way.  That's definitive enough for me :)

It's been a couple of months since it was exlpained but my understanding is that a DRBG in a chain (apart from the first) is not considered a RNG here.  I think the relevant section is NIST [SP 800-90C](https://csrc.nist.gov/CSRC/media/Publications/sp/800-90c/draft/documents/sp800_90c_second_draft.pdf) section 12.1.1 _Testing RBG Components_.  The point in question reads:

> The entropy source for the target RBG (or the initial RBG in the chain of RBGs) **shall** also be given a health test request as soon as it is available.

No subsequent DRBGs in a chain need to health test their input.  All of our DRBGs are in chains beginning from the master DRBG.  Only master requires the check and then only in FIPS mode.  If the tests are disabled, the seed state isn't stored -- by default (outside of FIPS mode), we'll be just as secure as before.

The FIPS documents don't stand in isolation and have to be understood holistically, hence I defer to the experts.

File: crypto/rand/drbg_lib.c [hunk 5]
```diff
@@ -324,10 +330,49 @@ void RAND_DRBG_free(RAND_DRBG *drbg)
...
+ * Perform the continuous random number generator (CRNG) test.
+ * Each block of entropy is compared against the previous and the test
+ * fails if any pair is identical.  Any excess bytes are zeroed and not used.
+ * The DRBG's old entropy block is updated so the test can continue on the
+ * next call.  The block size is assumed to be non-zero, i.e. the test is
+ * executing.
+ *
+ * Returns 1 on success, 0 on failure.
+ */
+static int crng_check(RAND_DRBG *drbg, size_t *lenp, unsigned char *entropy)
```
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/rand/drbg_lib.c]: Done

File: crypto/rand/drbg_lib.c (hunk 3 above)
Comment (paulidale) [reply crypto/rand/drbg_lib.c]: I'll have to ask about this one.

**Comment by mspncp** [comment]
> > Could you please explain why in your opinion the CRNG test needs to be applied only to the entropy _input_ of the 'master' DRBG and not to the _output_ generated by any of the DRBGs?
> 
> A good question. The short answer is that I've had confirmation from the FIPS lab and our internal FIPS guy that it can be done this way. That's definitive enough for me :)

You are probably right. I was just unsure about the precise meaning of the wording "employs ... RNGs"  in the FIPS document is:

> If a cryptographic module employs Approved or nonApproved RNGs ...

whether it applies to the FOM operating its own DRBG or to the FOM consuming the random input of a third party RNG. The FOM 2.0 took the former viewpoint, while you are taking the latter, which seems more plausible, given the fact that the FIPS guidelines where aimed primarily at hardware manufacturers.

**Comment by mspncp** [comment]
> It's been a couple of months since it was exlpained but my understanding is that a DRBG in a chain (apart from the first) is not considered a RNG here. I think the relevant section is NIST [SP 800-90C](https://csrc.nist.gov/CSRC/media/Publications/sp/800-90c/draft/documents/sp800_90c_second_draft.pdf) section 12.1.1 _Testing RBG Components_. The point in question reads:

>> The entropy source for the target RBG (or the initial RBG in the chain of RBGs) **shall** also be given a health test request as soon as it is available.

NIST SP 800-90B and 800-90C have detailed requirements for Health Tests of entropy sources, but those apply to the entropy source itself (e.g., the operating system's CSPRNG or a hardware TRNG), not to the DRBG consuming the entropy. Also, the tests are more sophisticated than the continous RNG test from FIPS 140-2, which is not mentioned at all.

Wouldn't it make sense to ask the officials whether the requirement of the continuous RNG test is still appropriate nowadays and might be dropped? After all, FIPS 140-2 was last updated in 2002, and the much newer NIST Special Publication 800-90B and 800-90C do not mention the test at all.  IMHO, testing for equality of consecutive blocks only is a very simplistic approach and only wastes processor cycles. (What about periodic input, almost constant input, ...?).

**Comment by paulidale** [comment]
I'm certain that the consecutive block test is still required, so we have to implement it.  I agree that it doesn't make a lot of sense for a software source.  NIST is quite hardware centric and from that viewpoint it makes more sense.

There is some recent draft guidance about using /dev/random from the operating system.  Essentially, it is permitted and doesn't require justification.  On the other hand, using /dev/urandom requires justification that it has been properly seeded (which is difficult).  I think we'll be safe if we insist on getentropy(2) or /dev/random as our sources.  The lab thinks that RDRAND & RDSEED could also be justified without problem.  This means a small change to our randomness collection but nothing significant.

**Comment by paulidale** [comment]
What's waiting for this one?

**Comment by paulidale** [comment]
Ping?

File: crypto/rand/drbg_lib.c [hunk 6]
```diff
@@ -380,16 +427,51 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
         max_entropylen += drbg->max_noncelen;
     }
 
+    /*
+     * If we're running with continuous RNG testing and haven't initialised
+     * yet, we will need more data.  We add one block and round up to the
+     * nearest block boundary.  The entropy requirements are not altered.
+     */
+    if (bs > 0) {
```
Comment (mspncp) [inline crypto/rand/drbg_lib.c]: To me it feels wrong and hacky to integrate the CRNG code into the DRBG implementation itself. In particular this tweaking of the input parameters gives me a bad feeling. In my opinion the design would be cleaner if the CRNG test would be implemented as a filter between the DRBG and its entropy source: It takes the request from the DRBG, requests the extra amount of random data from the source and tests the return data before handing it over to the DRBG. In other words, the testing should happen transparently to the DRBG.

Currently I am also pondering whether this extra block should always be discarded and not fed into the DRBG.

Comment (paulidale) [reply crypto/rand/drbg_lib.c]: NIST mandates the discard of the first block, so we'll need to be able to do it.
It does seem like a loss though.

**Comment by paulidale** [comment]
A filter seems reasonable.
Back to the drawing board.

File: crypto/rand/drbg_lib.c (hunk 3 above)
Comment (paulidale) [reply crypto/rand/drbg_lib.c]: I didn't note the response from the FIPS guy: the raw input needs to be used, the entropy requirement doesn't increase.

**Comment by paulidale** [comment]
closing this since it is obsolete.

**Pull request closed without merging**